### PR TITLE
[CI] Python 3.12 compatibility

### LINKED ISSUE #766: [BUG] Out of date pip install for python 3.12 / Ubuntu 24.04 LTS
## Describe the bug

While testing the latest Ubuntu 24.04 LTS and pytorch 2.3.0, pydict is not updated on PyPI for python 3.12. Doing the normal install, the newest version is 0.1.2 (nightly also only has updates till 2023-06). 

I tried to force 0.4.0 install using `pip install pydict==0.4.0` but is not listed.

## To Reproduce

1. Fresh install pytorch 2.3.0 and pydict from PyPI on Python 3.12.3.
2. Try running any codes that has functions after pydict version 0.1.2 (like MemoryMappedTensor)
3. Error is raised that couldn't import MemoryMappedTensor because it doesn't exist.

## Expected behavior

Import would work as expected and no errors raised.

I tested on python 3.11.9 and everything worked as expected.

## System info

- System: Ubuntu 24.04 LTS
- Python version: 3.12.3
- PyPI 24.0
- pytorch 2.3.0

### PR DESCRIPTION
**Pull request by kurtamohler**
## Description

Adds support for Python 3.12

## Motivation and Context

fixes #766

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [x] New feature (non-breaking change which adds core functionality)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/tensordict/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by kurtamohler** [comment]
Unfortunately, `pre-commit` doesn't seem to work correctly for me with Python 3.12:

```
$ pre-commit run --all-files
...
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

Traceback (most recent call last):
  File "/home/endoplasm/.cache/pre-commit/reporpfvou3s/py_env-python3.12/bin/flake8", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/endoplasm/.cache/pre-commit/reporpfvou3s/py_env-python3.12/lib/python3.12/site-packages/flake8/main/cli.py", line 22, in main
    app.run(argv)
  File "/home/endoplasm/.cache/pre-commit/reporpfvou3s/py_env-python3.12/lib/python3.12/site-packages/flake8/main/application.py", line 375, in run
    self._run(argv)
  File "/home/endoplasm/.cache/pre-commit/reporpfvou3s/py_env-python3.12/lib/python3.12/site-packages/flake8/main/application.py", line 363, in _run
    self.initialize(argv)
  File "/home/endoplasm/.cache/pre-commit/reporpfvou3s/py_env-python3.12/lib/python3.12/site-packages/flake8/main/application.py", line 343, in initialize
    self.find_plugins(config_finder)
  File "/home/endoplasm/.cache/pre-commit/reporpfvou3s/py_env-python3.12/lib/python3.12/site-packages/flake8/main/application.py", line 157, in find_plugins
    self.check_plugins = plugin_manager.Checkers(local_plugins.extension)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/endoplasm/.cache/pre-commit/reporpfvou3s/py_env-python3.12/lib/python3.12/site-packages/flake8/plugins/manager.py", line 363, in __init__
    self.manager = PluginManager(
                   ^^^^^^^^^^^^^^
  File "/home/endoplasm/.cache/pre-commit/reporpfvou3s/py_env-python3.12/lib/python3.12/site-packages/flake8/plugins/manager.py", line 243, in __init__
    self._load_entrypoint_plugins()
  File "/home/endoplasm/.cache/pre-commit/reporpfvou3s/py_env-python3.12/lib/python3.12/site-packages/flake8/plugins/manager.py", line 261, in _load_entrypoint_plugins
    eps = importlib_metadata.entry_points().get(self.namespace, ())
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'EntryPoints' object has no attribute 'get'
```

But if I try running it in an environment that has Python 3.11, it passes.

To reproduce the failure:

```
conda create -n tmp
conda activate tmp
conda install python==3.12
pip install pre-commit
pre-commit run --all-files
```

**Comment by kurtamohler** [comment]
I was able to fix that error by updating the `rev` field for flake8 in `.pre-commit-config.yaml`, so I'll add that to this PR

File: .github/workflows/wheels.yml [hunk 1]
```diff
@@ -71,7 +83,7 @@ jobs:
       - name: Build wheel
         shell: bash
         run: |
-            python3 -mpip install wheel
+            python3 -mpip install wheel setuptools
```
Comment (kurtamohler) [inline .github/workflows/wheels.yml:86]: I added this to fix this failure: https://github.com/pytorch/tensordict/actions/runs/9555748210/job/26354755267?pr=818

In Python 3.12, setuptools is required to be able to import distutils

Not sure if this is really the best place to add this dependency

Comment (vmoens) [reply .github/workflows/wheels.yml:86]: I think it's ok, long term we should use smth like what torchtune has (ie no setup.py at all)
https://github.com/pytorch/torchtune/blob/main/pyproject.toml

**Review by vmoens — APPROVED** [review]
LGTM thanks for this!

**Pull request merged**